x/liquidity/amm/pool.go

```diff
@@ -70,78 +80,240 @@ func (pool *BasicPool) IsDepleted() bool {
+// HighestBuyPrice returns the highest buy price of the pool.
+func (pool *BasicPool) HighestBuyPrice() (price sdk.Dec, found bool) {
+	// The highest buy price is actually a bit lower than pool price,
+	// but it's not important for our matching logic.
+	return pool.Price(), true
+}
 
-		// pc = floor(ps * min(x / rx, y / ry))
-		pc = ps.MulTruncate(sdk.MinDec(
-			x.ToDec().QuoTruncate(rx),
-			y.ToDec().QuoTruncate(ry),
-		)).TruncateInt()
+// LowestSellPrice returns the lowest sell price of the pool.
+func (pool *BasicPool) LowestSellPrice() (price sdk.Dec, found bool) {
+	// The lowest sell price is actually a bit higher than the pool price,
+	// but it's not important for our matching logic.
+	return pool.Price(), true
+}
 
-		mintProportion := pc.ToDec().Quo(ps)             // pc / ps
-		ax = rx.Mul(mintProportion).Ceil().TruncateInt() // ceil(rx * mintProportion)
-		ay = ry.Mul(mintProportion).Ceil().TruncateInt() // ceil(ry * mintProportion)
+func (pool *BasicPool) BuyAmountOver(price sdk.Dec, _ bool) (amt sdk.Int) {
+	if price.GTE(pool.Price()) {
+		return sdk.ZeroInt()
+	}
+	dx := pool.rx.ToDec().Sub(price.MulInt(pool.ry)).TruncateInt()
+	if dx.IsZero() {
+		return sdk.ZeroInt()
+	}
+	utils.SafeMath(func() {
+		amt = pool.rx.ToDec().QuoTruncate(price).Sub(pool.ry.ToDec()).TruncateInt()
+		if amt.GT(MaxCoinAmount) {
+			amt = MaxCoinAmount
+		}
 	}, func() {
-		ax, ay, pc = sdk.ZeroInt(), sdk.ZeroInt(), sdk.ZeroInt()
+		amt = MaxCoinAmount
 	})
-
 	return
 }
 
-// Withdraw returns withdrawn x and y coin amount when someone withdraws
-// pc pool coin.
-// Withdraw also takes care of the fee rate.
-func (pool *BasicPool) Withdraw(pc sdk.Int, feeRate sdk.Dec) (x, y sdk.Int) {
-	if pc.Equal(pool.ps) {
-		// Redeeming the last pool coin - give all remaining rx and ry.
-		x = pool.rx
-		y = pool.ry
-		return
+func (pool *BasicPool) SellAmountUnder(price sdk.Dec, _ bool) sdk.Int {
+	if price.LTE(pool.Price()) {
+		return sdk.ZeroInt()
 	}
+	return pool.ry.ToDec().Sub(pool.rx.ToDec().QuoRoundUp(price)).TruncateInt()
+}
 
+func (pool *BasicPool) BuyAmountTo(price sdk.Dec) (amt sdk.Int) {
+	if price.GTE(pool.Price()) {
+		return sdk.ZeroInt()
+	}
+	sqrtRx := utils.DecApproxSqrt(pool.rx.ToDec())
+	sqrtRy := utils.DecApproxSqrt(pool.ry.ToDec())
+	sqrtPrice := utils.DecApproxSqrt(price)
+	dx := pool.rx.ToDec().Sub(sqrtPrice.Mul(sqrtRx.Mul(sqrtRy))) // dx = rx - sqrt(P * rx * ry)
+	if dx.IsZero() {
+		return sdk.ZeroInt()
+	}
 	utils.SafeMath(func() {
-		proportion := pc.ToDec().QuoTruncate(pool.ps.ToDec())                             // pc / ps
-		multiplier := sdk.OneDec().Sub(feeRate)                                           // 1 - feeRate
-		x = pool.rx.ToDec().MulTruncate(proportion).MulTruncate(multiplier).TruncateInt() // floor(rx * proportion * multiplier)
-		y = pool.ry.ToDec().MulTruncate(proportion).MulTruncate(multiplier).TruncateInt() // floor(ry * proportion * multiplier)
+		amt = dx.QuoTruncate(price).TruncateInt() // dy = dx / P
```

Comment:
What if `price` is too small like `price=0`? Do we need some condition to set `price` as the lowest price?